### PR TITLE
Use git description for version number in startup banner

### PR DESCRIPTION
**Pull request by d-torrance**
As discussed in Zulip, we move away from the "X.Y.0.1" development numbers and instead use the more informative git description, e.g.:

```m2
Macaulay2, version 1.24.05-9-g936ba91285-dirty
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems,
               Isomorphism, LLLBases, MinimalPrimes, OnlineLookup,
               PrimaryDecomposition, ReesAlgebra, Saturation, TangentCone,
               Truncations, Varieties

i1 : 
```

We also:

* use `git describe` directly to create `version#"git desription"` so we're counting the number of commits since the last release tag rather than the number of commits since `M2/VERSION` was last touched (this new behavior was suggested in a TODO comment)
* modify `VERSION-README` with some updates to the release process (e.g., new calendar-based versioning scheme and using a PR rather than `pre-master`)
* update the copyright years in the `copyright` string

### CODE REVIEW

File: M2/Macaulay2/m2/startup.m2.in [hunk 1]
```diff
@@ -158,8 +158,9 @@ if firstTime then (
      toAbsolutePath = pth -> if pth =!= "stdio" and not isAbsolutePath pth then "/" | relativizeFilename("/", pth) else pth;
 
      copyright = (
-	  "Macaulay2, version " | version#"VERSION" | newline
-	  | "Copyright 2019 Daniel R. Grayson and Michael E. Stillman:" | newline
+	  "Macaulay2, version " |
+	  replace("^release-", "", version#"git description") | newline
+	  | "Copyright 1993-2024 Daniel R. Grayson and Michael E. Stillman:" | newline
```
Comment (mahrud) [inline M2/Macaulay2/m2/startup.m2.in:163]: Is the year strictly necessary here? (note that none of the libraries include year)

Also, might be easier to switch to `Macaulay2 release-1.YY.MM-...` so you don't need the regex replace.

File: M2/VERSION-README [hunk 2]
```diff
@@ -1,22 +1,18 @@
-
 The only place the Macaulay2 version number appears is in the file ./VERSION .
 
-We have three main branches on github:
+We have two main branches on github:
 
    development -- all pull requests are merged into this branch
-   pre-master  -- development is manually merged into this branch occasionally, triggering builds
-   master      -- pre-master is manually merged into this branch after builds succeed there.  Daily builds occcur here.
+   master      -- the most recent stable release
```
Comment (mahrud) [inline M2/VERSION-README:6]: I'd like to bring up the discussion of renaming `master` into `main`, `release`, or `stable` again, so you might want to hold off on this commit.

Comment (mahrud) [reply M2/VERSION-README:6]: Actually, this whole file probably should be deleted and move to a wiki page. Is there any reason to keep this in git?

File: M2/cmake/configure.cmake [hunk 3]
```diff
@@ -62,26 +62,14 @@ set(PACKAGE_VERSION ${Macaulay2_VERSION})
...
-    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
-    OUTPUT_VARIABLE   _release_commit)
-  execute_process(
-    COMMAND ${GIT_EXECUTABLE} rev-list ${_release_commit}..HEAD --count
-    ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE
-    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
-    OUTPUT_VARIABLE   COMMIT_COUNT)
-  set(GIT_DESCRIPTION version-${PROJECT_VERSION}-${COMMIT_COUNT}-${GIT_COMMIT})
+    OUTPUT_VARIABLE   GIT_DESCRIPTION)
+  set(GIT_DESCRIPTION ${GIT_DESCRIPTION})
```
Comment (mahrud) [inline M2/cmake/configure.cmake:69]: I don't think this last line is required.

Comment (mahrud) [inline M2/cmake/configure.cmake:69]: This line is not required.
```suggestion
```

**Review by mahrud — CHANGES_REQUESTED** [review]

File: M2/VERSION-README (hunk 2 above)
Comment (d-torrance) [reply M2/VERSION-README:6]: Probably not. There are at least 4 different places where various parts of the release process are documented, so combining them into just a single wiki page seems like a good idea!

Comment (mahrud) [reply M2/VERSION-README:6]: Let's delete this then. Where are the other 3?

File: M2/Macaulay2/m2/startup.m2.in (hunk 1 above)
Comment (d-torrance) [reply M2/Macaulay2/m2/startup.m2.in:163]: All the GPL documentation says to include the year, e.g. from https://www.gnu.org/licenses/gpl-howto.en.html

> The copyright notice should include the year in which you finished preparing the release

I'm not sure why we include all the libraries in the notice. I haven't seen that before in other projects

File: M2/VERSION-README (hunk 2 above)
Comment (d-torrance) [reply M2/VERSION-README:6]: This file, the wiki page, `INSTALL`, and a one of the `bugs/dan` files

File: M2/Macaulay2/m2/startup.m2.in (hunk 1 above)
Comment (mahrud) [reply M2/Macaulay2/m2/startup.m2.in:163]: So we need to change that line every year? That seems tedious (we could also infer it from the new version scheme, but that seems over engineered ...)

I don't even understand why this has to be a part of startup.m2! A documentation page seems much more appropriate.

File: M2/cmake/configure.cmake (hunk 3 above)
Comment (d-torrance) [reply M2/cmake/configure.cmake]: Done!

File: M2/VERSION-README (hunk 2 above)
Comment (d-torrance) [reply M2/VERSION-README]: The [wiki page](https://github.com/Macaulay2/M2/wiki/Internals%3A-Release-Checklist) has been updated to contain the (updated) contents of the removed files/sections.

**Comment by d-torrance** [comment]
Based on the latest Zulip discussions, I've added the branch information (unless we're at a release tag):

```m2
Macaulay2, version 1.24.05-14-g2cfba68d03-dirty (version)
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems,
               Isomorphism, LLLBases, MinimalPrimes, OnlineLookup,
               PrimaryDecomposition, ReesAlgebra, Saturation, TangentCone,
               Truncations, Varieties

i1 :
```

**Review by mahrud — APPROVED** [review]

File: M2/Macaulay2/m2/startup.m2.in (hunk 1 above)
Comment (mahrud) [reply M2/Macaulay2/m2/startup.m2.in:167]: Here's the full suggestion: delete that stuff from startup.m2 and make `copyright` be a command that opens `help "Copyright and license"`. Also delete the documentation page "copyright".

Comment (d-torrance) [reply M2/Macaulay2/m2/startup.m2.in]: Done!

I also added a short copyright blurb to the startup banner that we probably should have had all along as per the GPL:

> If the program is interactive, make it output a short notice like this when it starts in an interactive mode:
>
>   ```
>   Gnomovision version 69, Copyright (C) year name of author
>   Gnomovision comes with ABSOLUTELY NO WARRANTY; for details type `show w'.
>   This is free software, and you are welcome to redistribute it
>   under certain conditions; type `show c' for details.
>  ```

R and Octave, both of which are also GPL, both have something like this.  So this is the new proposal:

```m2
Macaulay2, version 1.24.05-14-g5d5a2de127-dirty (version)
Copyright 1993-2024 Daniel R. Grayson and Michael E. Stillman
This program comes with ABSOLUTELY NO WARRANTY.  This is free
software, and you are welcome to redistribute it under certain
conditions; type 'copyright' for details.
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems,
               Isomorphism, LLLBases, MinimalPrimes, OnlineLookup,
               PrimaryDecomposition, ReesAlgebra, Saturation, TangentCone,
               Truncations, Varieties

i1 : copyright

o1 = Copyright and license
     *********************

     Macaulay2, its object code, source code, and documentation, are copyright
     by Daniel R. Grayson and Michael E. Stillman.  We permit you to use it
     either under the terms of the GNU General Public License, version 2, as
     published by the Free Software Foundation, and as contained in the file
     "COPYING-GPL-2" accompanying the program, or under the terms of the GNU
     General Public License, version 3, as published by the Free Software
     Foundation, and as contained in the file "COPYING-GPL-3" accompanying the
     program.

     Various packages are included with Macaulay2 and come with their own
     copyright notices and licenses, see "packages provided with Macaulay2".

     Some free libraries have been compiled into (or linked with) Macaulay2, and
     some free programs, or packages of programs, with their libraries, have
     been compiled and distributed with Macaulay2:

     Menu
     ----

     libraries

       * "Singular-Factory"
       * "frobby"
       * "MPIR"
       * "GNU MP"
       * "MPFR"
       * "MPFI"
       * "GC garbage collector"
       * "LAPACK"
       * "BLAS"
       * "NTL library"
       * "FLINT"
       * "givaro"
       * "fflas-ffpack"
       * "MPSolve" -- a library for finding roots of univariate polynomials
       * "Boost" -- a collection of C++ source libraries

     programs and their libraries

       * "4ti2"
       * "gfan"
       * "normaliz"
       * "csdp"
       * "nauty"
       * "cdd+"
       * "lrslib"
       * "topcom"
       * "cohomCalg"

```
The `--copyright` command line option now just adds this new command to the list of startup functions using `addStartFunction`.

Comment (mahrud) [reply M2/Macaulay2/m2/startup.m2.in]: Hmm I really dislike this actually lol

Comment (mahrud) [reply M2/Macaulay2/m2/startup.m2.in]: If we must, can we either:
1. condition it to display only on the first run (or if `-q` is given)
2. make it just mention the commands like python:
```
Type "help", "copyright", "credits" or "license" for more information.
```
Admittedly they are not using GPL, but we haven't had this notice until now and nothing has gone wrong, I doubt someone will steal M2 if we don't remind them of the copyright every time it runs XD

**Comment by mahrud** [comment]
This wasn't ready yet.

**Comment by mahrud** [comment]
This is way too much unnecessary banner as the default (i.e. without the `--copyright` option)
```m2
Macaulay2, version 1.24.05-37-gf5d39b2090 ()
Copyright 1993-2024 Daniel R. Grayson and Michael E. Stillman
This program comes with ABSOLUTELY NO WARRANTY.  This is free
software, and you are welcome to redistribute it under certain
conditions; type 'copyright' for details.
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems,
               Isomorphism, LLLBases, MinimalPrimes, OnlineLookup,
               PrimaryDecomposition, ReesAlgebra, Saturation, TangentCone,
               Truncations, Varieties
```
Also the branch seems to be empty.

**Comment by mahrud** [comment]
Couple of more point:
1. [Per GPL](https://www.gnu.org/licenses/gpl-howto.en.html), for interactive programs "it is **usually a good idea** to make the program display a brief notice about copyright and copying permission when it starts up," so it is not a requirement.
2. As an example, Bash is GPLv3 licensed but doesn't blast everyone with a copyright notice unless you use `bash --version`.

**Comment by mikestillman** [comment]
I agree with @mahrud , I would like the default to be a very short banner.

File: M2/Macaulay2/d/CMakeLists.txt [hunk 4]
```diff
@@ -135,7 +135,8 @@ endforeach()
 
 # To avoid unnecessary recompiling, only version.dd is given GIT_DESCRIPTION
 set_source_files_properties(version-tmp.cc
-  PROPERTIES COMPILE_DEFINITIONS GIT_DESCRIPTION="${GIT_DESCRIPTION}")
+  PROPERTIES COMPILE_DEFINITIONS
+  "GIT_DESCRIPTION=\"${GIT_DESCRIPTION}\";GIT_BRANCH=\"${GIT_BRANCH}\"")
```
Comment (mahrud) [inline M2/Macaulay2/d/CMakeLists.txt:139]: This is a little awkward, I would prefer:
```suggestion
set_property(SOURCE version-tmp.cc APPEND PROPERTY COMPILE_DEFINITIONS
  GIT_DESCRIPTION="${GIT_DESCRIPTION}"
  GIT_BRANCH="${GIT_BRANCH}")
```

File: M2/cmake/configure.cmake [hunk 5]
```diff
@@ -62,26 +62,19 @@ set(PACKAGE_VERSION ${Macaulay2_VERSION})
...
+    COMMAND ${GIT_EXECUTABLE} describe --tags --dirty
     ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE
     WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
-    OUTPUT_VARIABLE   GIT_COMMIT)
-  # TODO: currently finds the last commit that changed the VERSION file
-  # but ideally it should get the last release commit instead
+    OUTPUT_VARIABLE   GIT_DESCRIPTION)
   execute_process(
-    COMMAND ${GIT_EXECUTABLE} rev-list -1 HEAD VERSION
+    COMMAND ${GIT_EXECUTABLE} branch --show-current
```
Comment (mahrud) [inline M2/cmake/configure.cmake:70]: Turns out if you're on a detached state (say if you checkout `upstream/development` or someone else's branch) this is just an empty string, which shows up as `()` in the version string. An alternative might be `git show -s --pretty=%D HEAD` but it's a bit verbose. Here are a couple of examples:
```
HEAD -> SheafMaps, origin/SheafMaps
HEAD, tag: release-1.24.05, upstream/release-1.24.05-branch, upstream/master
```

Comment (d-torrance) [reply M2/cmake/configure.cmake:70]: If we're not on a branch, then getting the empty string for `version#"git branch"` seems like reasonable behavior.  We have `version#"git description"` to tell us more already.

Maybe just don't include the branch info in the startup banner if it's empty?

File: M2/Macaulay2/d/CMakeLists.txt (hunk 4 above)
Comment (d-torrance) [reply M2/Macaulay2/d/CMakeLists.txt:139]: Thanks for the suggestion!  Applied in https://github.com/Macaulay2/M2/pull/3247/commits/b4236cfeb488f4fe908bc4b179c8d8433f8df4d9

File: M2/cmake/configure.cmake (hunk 5 above)
Comment (mahrud) [reply M2/cmake/configure.cmake:70]: That seems reasonable.

Also, I realized a major drawback of including anything more than just the version in the binary is that any change in git will cause M2-binary to recompile. It's not urgent to change this, but I think a possible solution is to move the most frequently changed numbers into a non-compiled file (e.g. a variable set in an m2 file).

**Pull request merged**